compare: ufs-fire integration test changes

benkozi opened the pull request.

benkozi committed:
initial ufs-fire test

benkozi committed:
initial ufs-fire test

benkozi committed:
initial ufs-fire test

EdwardSnyder-NOAA committed:
[develop] Add NOAA Cloud coverage tests and skill-score WE2E to all coverage suites (#1200)

* Adds NOAA Cloud coverage test suites for each cloud service providers
* Adds the skill-score WE2E test (grid_SUBCONUS_Ind_3km_ics_FV3GFS_lbcs_FV3GFS_suite_WoFS_v0) to each coverage suite via the Jenkinsfile

---------

Co-authored-by: michael.lueken <Michael.Lueken@noaa.gov>

benkozi committed:
initial ufs-fire test

benkozi committed:
switching branches

benkozi committed:
add configuration parameter to yaml

benkozi committed:
add to e2e tests

benkozi commented:
@mkavulich - When you have a minute, can you peek at these initial changes adding UFS-Fire file checking to the integration test task? I'm happy to add more in-depth tests/configuration if needed. I didn't want to go too crazy until I knew what was useful, and if this type of test is in fact what you had in mind.

benkozi committed:
add to e2e tests

benkozi committed:
fix flag

mkavulich commented on jobs/JREGIONAL_INTEGRATION_TEST:
```diff
@@ -98,9 +99,15 @@ fi
 #
 #-----------------------------------------------------------------------
 #
+if [ $(boolify "${TEST_UFS_FIRE}") = "TRUE" ]; then
```
Since uwtools is now used to read in all variables from the config file, we shouldn't need this "boolify" function anymore since the string representation of booleans in bash will always be the same.

benkozi replied:
Oh! I didn't know that. Is it always `TRUE` or `FALSE`?

benkozi replied:
Ahhh...`True`/`False`

mkavulich replied:
Yes, the strings read in should now always be the pythonic `True` or `False`. I just opened an issue about this (#1280); every bit of simplification helps in my opinion!

mkavulich commented on ush/config.fire.yaml:
```diff
@@ -63,3 +63,6 @@ fire:
   FIRE_UPWINDING: 9
   FIRE_LSM_ZCOUPLING: False
   FIRE_LSM_ZCOUPLING_REF: 60.0
+task_integration_test:
```
I don't think the integration test should be included in the user example file, this could be confusing.

benkozi replied:
Done!

mkavulich reviewed (commented):
@benkozi This is a good start! One additional feature I'd suggest is checking that the correct number of fire output files is present, not just any number greater than zero. You could do this with some math on `FCST_LEN_HRS` and `OUTPUT_DT_FIRE`.

Another useful feature would be checking that the fire.namelist is rendered properly, but that can be further down the road as it would require more work for sure.

Also, I know this is outside of the scope of this PR, but I just realized we don't actually test differing OUTPUT_DT_FIRE values... could you modify `config.UFS_FIRE_multifire_two-way-coupled.yaml` to use `OUTPUT_DT_FIRE: 600`? If for some reason that causes errors let me know and I can take care of it.

benkozi commented:
> @benkozi This is a good start! One additional feature I'd suggest is checking that the correct number of fire output files is present, not just any number greater than zero. You could do this with some math on `FCST_LEN_HRS` and `OUTPUT_DT_FIRE`.
>
> Another useful feature would be checking that the fire.namelist is rendered properly, but that can be further down the road as it would require more work for sure.

Sure, let me look into it. I think with this type of calculation in the test, it would be best to read these values from the namelist directly. It could quickly turn into a bit of a mess lifting things up as `envvar`s. Another option would be to pass the expected number of files as an argument. I like the direct-from-namelist approach the best though.

> Also, I know this is outside of the scope of this PR, but I just realized we don't actually test differing OUTPUT_DT_FIRE values... could you modify `config.UFS_FIRE_multifire_two-way-coupled.yaml` to use `OUTPUT_DT_FIRE: 600`? If for some reason that causes errors let me know and I can take care of it.

Will do!

benkozi committed:
address comments and other improvements (#26)

benkozi commented on scripts/exregional_integration_test.py:
```diff
@@ -27,34 +27,128 @@
+            raise ValueError
+        return self._namelist_fire
+
+    def test_output_files_created(self) -> None:
+        cfg = self.get_config()
+        fire_files = tuple(cfg.fcst_dir.glob("*fire_output_*nc"))
+        n_actual_files = len(fire_files)
+        logging.info(f"{n_actual_files=}")
+        interval_output = self._namelist_fire["time"]["interval_output"]
+        n_expected_files = int(((cfg.fcst_len * 60 * 60) / interval_output) + 1)
+        logging.info(f"{n_expected_files=}")
+        self.assertEqual(n_actual_files, n_expected_files)
+
+    def test_namelist_created(self) -> None:
+        expected_keys = {'time': ('dt', 'interval_output'), 'atm': ('interval_atm', 'kde'),
+                         'fire': ('fire_num_ignitions', 'fire_ignition_ros1',
+                                  'fire_ignition_start_lat1', 'fire_ignition_start_lon1',
+                                  'fire_ignition_end_lat1', 'fire_ignition_end_lon1',
+                                  'fire_ignition_radius1', 'fire_ignition_start_time1',
+                                  'fire_ignition_end_time1', 'fire_wind_height',
+                                  'fire_print_msg', 'fire_atm_feedback', 'fire_viscosity',
+                                  'fire_upwinding', 'fire_lsm_zcoupling',
+                                  'fire_lsm_zcoupling_ref')}
+
+        namelist_fire = self.get_namelist_fire()
+        self.assertEqual(set(namelist_fire.keys()), set(expected_keys.keys()))
+        for key in expected_keys.keys():
+            expected_group_keys = set(expected_keys[key])
+            actual_group_keys = set(namelist_fire[key].keys())
+            # There can be multiple entries for keys suffixed with "1". We are not testing multiple
+            # parameter entries here.
+            self.assertTrue(expected_group_keys.issubset(actual_group_keys))
+            logging.info(f"{actual_group_keys.difference(expected_group_keys)=}")
```
@mkavulich - Let me know what you think of these test changes. The we2e tests pass with these changes and the 600 second history interval for the multi-fire test.

mkavulich replied:
```suggestion
    def test_output_files_created(self) -> None:
        cfg = self.get_config()
        fire_files = tuple(cfg.fcst_dir.glob("*fire_output_*nc"))
        n_actual_files = len(fire_files)
        logging.info(f"{n_actual_files=}")
        interval_output = self._namelist_fire["time"]["interval_output"]
        n_expected_files = int(((cfg.fcst_len * 60 * 60) / interval_output) + 1)
        logging.info(f"{n_expected_files=}")
        self.assertEqual(n_actual_files, n_expected_files)

    def test_namelist_created(self) -> None:
        expected_keys = {'time': ['dt', 'interval_output'], 'atm': ['interval_atm', 'kde'],
                         'fire': ['fire_num_ignitions', 'fire_wind_height',
                                  'fire_print_msg', 'fire_atm_feedback', 'fire_viscosity',
                                  'fire_upwinding', 'fire_lsm_zcoupling',
                                  'fire_lsm_zcoupling_ref']}
        multifire_params = ['fire_ignition_ros', 'fire_ignition_start_lat',
                            'fire_ignition_start_lon', 'fire_ignition_end_lat',
                            'fire_ignition_end_lon', 'fire_ignition_radius',
                            'fire_ignition_start_time', 'fire_ignition_end_time' ]


        namelist_fire = self.get_namelist_fire()
        num_fires = namelist_fire['fire']['fire_num_ignitions']
        # For each fire we need one of these settings in the namelist with an integer suffix
        for i in range(num_fires):
            for param in multifire_params:
                expected_keys['fire'].append(param + str(i+1))

        self.assertEqual(set(namelist_fire.keys()), set(expected_keys.keys()))
        for key in expected_keys.keys():
            expected_group_keys = set(expected_keys[key])
            actual_group_keys = set(namelist_fire[key].keys())
            self.assertTrue(actual_group_keys.issubset(expected_group_keys))
            logging.info(f"{expected_group_keys.difference(actual_group_keys)=}")
```

This logic will ensure that all namelist options (including those dependent on the number of fires) are present. Note that I switched the check from the expected being a subset of the actual to the actual being a subset of the expected. While there are options in the fire model that aren't currently tested (and so would be valid namelist options not listed here) we can add that to the integration check in the future as needed.

mkavulich replied:
Somehow my initial suggestion got messed up in the diff, I think it's fixed now.

benkozi replied:
Nice, I had thought we'd have to check the config file to get number of ignitions. I should have scrubbed the namelist more closely.

I incorporated your changes. Resolving

mkavulich commented on scripts/exregional_integration_test.py:
```diff
@@ -27,34 +27,128 @@
+    _namelist_fire: f90nml.Namelist | None = None
+
+    @classmethod
+    def setUpClass(cls) -> None:
```
`uwtools` has standard handling of namelists that I think we should be leveraging here rather than creating our own read functions with `f90nml`. It might not simplify this code a ton but I think keeping our handling consistent will make things easier in the long run.

https://uwtools.readthedocs.io/en/main/sections/user_guide/api/config.html#uwtools.api.config.get_nml_config

https://uwtools.readthedocs.io/en/main/sections/user_guide/api/config.html#uwtools.api.config.NMLConfig

benkozi replied:
Oh nice, I wasn't aware. I switched over to `uwtools` and also rearranged the logger.

benkozi committed:
use uwtools, rearrange logging (#27)

benkozi committed:
add multifire testing support (#28)

benkozi committed:
[develop] feat: history files can be written to FV3 native grid (#1273)

mkavulich committed:
[develop] Add checks for correct SPP settings for the selected CCPP suite, resolve four more issues (#1259)

This PR adds a check in setup.py that ensures the appropriate physics schemes are being used for the selected SPP settings, and if not, removes the SPP settings for that scheme. If after this pruning there are no SPP settings left, an exception is raised. Note that many of the changes in setup.py are just re-arranging of existing code; the removed logic from lines 1029 to 1141 are moved downwards so that all of these CCPP checks are done only when a run_fcst task is specified in the workflow.

While making these changes to setup.py, I also added this file to the list of pytest-ed files. This required updating pylint to a newer version, since 2.17 results in an error (and this older version is no longer supported anyway).

Additionally, this PR resolves four more issues:
* Issue #1228, a new smoke/dust test smoke_dust_grid_RRFS_CONUS_25km_suite_RRFS_sas is added
* Issue #1258, removing smoke/dust entries from parm/field_table.RRFS_sas
* Issue #1275, fixing the order of operations error in setup.py after issue #1204 was merged
* Issue #1276, modified WE2E test get_from_HPSS_ics_GDAS_lbcs_GDAS_fmt_netcdf_2022040400_ensemble to test SPP_LSM, also change grid from CONUS 3km to CONUS 13km to make it much cheaper (we already have plenty CONUS 3km tests)

---------

Co-authored-by: Ben Koziol <bkoziol@redlineperf.com>
Co-authored-by: michael.lueken <Michael.Lueken@noaa.gov>

benkozi committed:
[develop] fix: adjust fire config for uwtools integration (#1278)

benkozi committed:
Merge remote-tracking branch 'ufs-srweather/develop' into feat/d1215-fire-output-validation

benkozi commented:
Closing this and will reopen to SRW develop. Thank you for the feedback here. I'll link to this for reference.
